db/system_keyspace.cc

```diff
@@ -964,6 +966,24 @@ schema_ptr system_keyspace::broadcast_kv_store() {
     return schema;
 }
 
+schema_ptr system_keyspace::sstables_registry() {
+    static thread_local auto schema = [] {
+        auto id = generate_legacy_id(NAME, SSTABLES_REGISTRY);
+        return schema_builder(NAME, SSTABLES_REGISTRY, id)
+            .with_column("location", utf8_type, column_kind::partition_key)
+            .with_column("generation", long_type, column_kind::clustering_key)
+            .with_column("uuid", uuid_type)
```

Comment:
why both generation and uuid?

with uuid based generation, it seems having both generation and uuid is redundant.

additionally, we can have collision with 64-bit generation being used as clustering key.

even if we use uuid, using generation_type::value() means we're losing half of the info from the 128-bit uuid.

I suggest using uuid_type for generation, and refusing integer based generation with s3 storage.

Reply:
> why both generation and uuid?

generation because it's still there and it's needed by the rest of sstables code
uuid is here because I need a unique identifier for object on the s3 side

> with uuid based generation, it seems having both generation and uuid is redundant.

Of course. But as long as it's not yet there, we have to stick with both

> additionally, we can have collision with 64-bit generation being used as clustering key.

Hm... Nowadays 64-bit generation is used as a "unique" identifier for files on /var/lib/data directory, so having a conflict as clustering key means we can have conflict on filesystem too

> even if we use uuid, using generation_type::value() means we're losing half of the info from the 128-bit uuid.
> 
> I suggest using uuid_type for generation, and refusing integer based generation with s3 storage.

Definitely. Once uuid generations are in we'll do it

Reply:
The uuid field is 100% auxiliary here. It's only purpose it to provide S3-side unique object name. Once uuid-generation gets merged this field will be just removed from the table. I'm all for simplifying this part right now, but I don't see a way to shoehorn generation into a uuid just because generation is the clustering key, and it's essential to be such, and uuid is regular column which cannot be part of the clustering key, because it's the value the code needs to resolve with `location:generation` pair at hand.

Maybe an option to lower the WTF level around this thing is not to call it uuid and not to generate it as uuid? But use smth like random::get_random_string() instead? This would more-or-less work for "s3-side uniqueness" temporarily.

Reply:
> > why both generation and uuid?
> 
> generation because it's still there and it's needed by the rest of sstables code uuid is here because I need a unique identifier for object on the s3 side
> 
> > with uuid based generation, it seems having both generation and uuid is redundant.
> 
> Of course. But as long as it's not yet there, we have to stick with both
> 
> > additionally, we can have collision with 64-bit generation being used as clustering key.
> 
> Hm... Nowadays 64-bit generation is used as a "unique" identifier for files on /var/lib/data directory, so having a conflict as clustering key means we can have conflict on filesystem too
> 

I was thinking ahead to shared storage, where 64-bit generation can collide across nodes in the same cluster.

So the idea is that we can modify the schema at time we want as the feature is not GA yet. Got it. thanks for the explanation.

Reply:
Ah skipped this in the cover letter: "maintains the system.sstables_registry table to keep track of what objects belong to local node and their names." :-) sorry for the noise.